File: dcp-solutions/2024-12-22/solution.js
```javascript
function shortestSubstring(s, chars) {
    if (!s || !chars.length) {
        return "";
    }

    let charMap = {};
    for (let char of chars) {
        charMap[char] = 0;
    }

    let left = 0;
    let minLen = Infinity;
    let minStr = "";

    let formed = 0;
    for (let right = 0; right < s.length; right++) {
        let char = s[right];
        if (char in charMap) {
            charMap[char]++;
            if (charMap[char] === 1) {
                formed++;
            }
        }

        while (formed === chars.length && left <= right) {
            if (right - left + 1 < minLen) {
                minLen = right - left + 1;
                minStr = s.slice(left, right + 1);
            }

            let char = s[left];
            charMap[char]--;
            if (char in charMap && charMap[char] === 0) {
                formed--;
            }
            left++;
        }
    }

    return minStr;
}
```

File: dcp-solutions/2024-12-22/solution.ex
```elixir
defmodule ShortestSubstring do
  @doc """
  Returns the shortest substring of `s` containing all characters in `chars`.

  Args:
    s (String.t): The string to search.
    chars (Set.t): A set of characters.

  Returns:
    String.t: The shortest substring or nil if no such substring exists.
  """
  def find_shortest_substring(s, chars) do
    # If the set of characters is empty, return an empty string
    if Enum.empty?(chars), do: ""

    # Create a map to store the frequency of each character in `chars`
    char_freq = Map.new()
    for char <- chars do
      char_freq[char] = 0
    end

    # Initialize variables to keep track of the minimum substring
    min_len = Float.inf
    min_substring = nil

    # Initialize variables to keep track of the current window
    left = 0
    current_chars = %{}

    # Iterate over each character in `s`
    for right in 1..length(s) do
      char = String.slice(s, right - 1)

      # If the character is in `chars` and it's not already in the window
      if Map.get(char_freq, char) > 0 && !(Map.get(current_chars, char)) do
        # Increment its frequency
        char_freq[char] = char_freq[char] + 1

        # Add it to the window
        current_chars[char] = true
      end

      # If all characters in `chars` are in the window
      if Enum.all?(for <<char, _ >> <- Map.keys(char_freq), into: &(Map.get(&1, char) > 0)), do: true, else
        # Move the left pointer to the right
        left = left + 1

        # Decrement the frequencies of characters going out of the window
        for char <- current_chars do
          case Map.get(char_freq, char) do
            0 -> Map.delete(char_freq, char)
            _ when Map.get(char_freq, char) > 1 -> Map.put(char_freq, char, Map.get(char_freq, char) - 1)
          end
        end

        # Update the minimum substring if necessary
        length = right - left + 1
        if length < min_len do
          min_len = length
          min_substring = String.slice(s, left, right)
        end
      end
    end

    # If a valid substring was found, return it; otherwise, return nil
    case min_substring do
      nil -> nil
      _ when min_len == Float.inf -> nil
      ^min_substring
    end
  end
end
```